accounts-db/src/accounts_db.rs

```diff
@@ -7819,18 +7819,45 @@ impl AccountsDb {
 
         let (dirty_pubkeys, insert_time_us, mut generate_index_results) = {
             let mut items_local = Vec::default();
-            storage.accounts.scan_index(|info| {
-                stored_size_alive += info.stored_size_aligned;
-                if info.index_info.lamports > 0 {
-                    accounts_data_len += info.index_info.data_len;
-                    all_accounts_are_zero_lamports = false;
-                } else {
-                    // zero lamport accounts
-                    zero_lamport_pubkeys.push(info.index_info.pubkey);
-                }
-                items_local.push(info.index_info);
-            });
+            if secondary {
+                // WITH secondary indexes -- scan accounts WITH account data
+                storage.accounts.scan_accounts(|offset, account| {
+                    let data_len = account.data.len() as u64;
+                    stored_size_alive += storage.accounts.calculate_stored_size(data_len as usize);
+                    if account.lamports > 0 {
+                        accounts_data_len += data_len;
+                        all_accounts_are_zero_lamports = false;
+                    } else {
+                        // zero lamport accounts
+                        zero_lamport_pubkeys.push(*account.pubkey);
+                    }
+                    items_local.push(IndexInfoInner {
+                        offset,
+                        pubkey: *account.pubkey,
+                        lamports: account.lamports,
+                        data_len,
+                    });
 
+                    self.accounts_index.update_secondary_indexes(
+                        account.pubkey,
+                        &account,
+                        &self.account_indexes,
+                    );
```

Comment:
why replicode instead of wrapping this block in `if secondary { }`. seems no else is necessary

Reply:
where does the top of this closure diverge in equivalence from the closure in the else statement. is there a perf hit under the hood? looks like a minor adaptation to different inputs from here

Reply:
i can get on board with this version